stan/math/rev/core/arena_allocator.hpp

```diff
@@ -13,33 +13,68 @@ namespace math {
 template <typename T>
 struct arena_allocator {
   using value_type = T;
+  using pointer = T*;
+  using const_pointer = const T*;
+  using reference = T&;
+  using const_reference = const T&;
+  using size_type = size_t;
+  using difference_type = ptrdiff_t;
+  using propagate_on_container_move_assignment = std::true_type;
+
+  constexpr arena_allocator() noexcept {};
+
+  template <class U>
+  constexpr arena_allocator(const arena_allocator<U>&) noexcept {}
+
+  template <class U>
+  constexpr arena_allocator<T>& operator=(const arena_allocator<U>&) noexcept {
+    return *this;
+  }
+
+  static constexpr size_type max_size() noexcept {
+    return std::numeric_limits<size_type>::max() / sizeof(T);
+  }
 
   /**
    * Allocates space for `n` items of type `T`.
    *
    * @param n number of items to allocate space for
    * @return pointer to allocated space
    */
-  T* allocate(std::size_t n) {
+  static inline T* allocate(std::size_t n) noexcept {
     return ChainableStack::instance_->memalloc_.alloc_array<T>(n);
   }
 
   /**
    * No-op. Memory is dealocated by caling `recover_memory()`.
    */
-  void deallocate(T* /*p*/, std::size_t /*n*/) noexcept {}
+  inline void deallocate(T* /*p*/, std::size_t /*n*/) const noexcept {}
+
+  static inline void construct(pointer p, const_reference val) {
+    new (static_cast<void*>(p)) T(val);
+  }
+
+  static inline void destroy(pointer p) { (static_cast<T*>(p))->~T(); }
```

Comment:
No need for cast. p is of type T*. I would prefere if you used `T*` / `T&` everywhere instead of typedefs, so such mistakes would not happen.